src/OpenLoco/Map/MapGenerator.cpp

```diff
@@ -427,14 +430,199 @@ namespace OpenLoco::Map::MapGenerator
+        {
+            return std::nullopt;
+        }
+        if (surface.hasHighTypeFlag())
+        {
+            return std::nullopt;
+        }
+
+        auto* landObj = ObjectManager::get<LandObject>(surface.terrain());
+        if (landObj == nullptr)
+        {
+            return std::nullopt;
+        }
+
+        if (landObj->var_1B == 0 || surface.slope())
+        {
+            return 0;
+        }
+
+        // TODO: split into two randNext calls
+        uint16_t randVal = gPrng().randNext();
+        if (landObj->var_1C <= (randVal >> 8))
+        {
+            return 0;
+        }
+        return ((randVal & 0xFF) * landObj->var_1B) >> 8;
+    }
+
+    // 0x0046A379
+    static void generateTerrainFarFromWater(HeightMap& heightMap, uint8_t surfaceStyle)
+    {
+        _heightMap = heightMap.data();
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A379, regs);
+        _heightMap = nullptr;
+    }
+
+    // 0x0046A439
+    static void generateTerrainNearWater(HeightMap& heightMap, uint8_t surfaceStyle)
+    {
+        _heightMap = heightMap.data();
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A439, regs);
+        _heightMap = nullptr;
+    }
+
+    // 0x0046A5B3
+    static void generateTerrainOnMountains(HeightMap& heightMap, uint8_t surfaceStyle)
+    {
+        _heightMap = heightMap.data();
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A5B3, regs);
+        _heightMap = nullptr;
+    }
+
+    // 0x0046A4F9
+    static void generateTerrainFarFromMountains(HeightMap& heightMap, uint8_t surfaceStyle)
+    {
+        _heightMap = heightMap.data();
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A4F9, regs);
+        _heightMap = nullptr;
+    }
+
+    // 0x0046A0D8
+    static void generateTerrainInSmallRandomAreas(HeightMap& heightMap, uint8_t surfaceStyle)
     {
         _heightMap = heightMap.data();
-        call(0x0046A021);
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A0D8, regs);
         _heightMap = nullptr;
     }
 
+    // 0x0046A227
+    static void generateTerrainInLargeRandomAreas(HeightMap& heightMap, uint8_t surfaceStyle)
+    {
+        _heightMap = heightMap.data();
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A227, regs);
+        _heightMap = nullptr;
+    }
+
+    // 0x0046A66D
+    static void generateTerrainAroundCliffs(HeightMap& heightMap, uint8_t surfaceStyle)
+    {
+        _heightMap = heightMap.data();
+        registers regs;
+        regs.ebx = surfaceStyle;
+        call(0x0046A66D, regs);
+        _heightMap = nullptr;
+    }
+
+    static void generateTerrainNull(HeightMap& heightMap, uint8_t surfaceStyle) {}
+
+    using GenerateTerrainFunc = void (*)(HeightMap&, uint8_t);
+    static const GenerateTerrainFunc _generateFuncs[] = {
+        generateTerrainNull,
+        generateTerrainNull,
```

Comment:
any reason for the 2 'generate null' functions?
edit: after reading further, there should definitely only be 1 'generate null' here
edit2: after even further reading, I see the first 2 nulls are for the first 2 entries in the enum `LandDistribuitionPattern` (which is mispelt, but that's another issue)